File: spikes/treebuild.py
import sys
import json
import uuid

if __name__ == '__main__':
    with open(sys.argv[1]) as f:
        doc = json.load(f)

    def build_tree(top):
        top["children"] = [child for child in doc if child["parent"] == top["id"]]
        for child in top["children"]:
            build_tree(child)
        return top

    root = [x for x in doc if x['parent'] == None][0]
    print build_tree(root)

File: spikes/treebuild.js
var db = require("./flatcouch.json"),
    _ = require("lodash"),
    root = _.find(db, function (x) { return _.isNull(x.parent); });

function build_tree(top) {
    return _.merge({}, top, {
        children: _.map(_.filter(db, function (x) { return x.parent === top.id}), build_tree)
    });
}

console.log(JSON.stringify(build_tree(root), null, "  "));